node/ChainwebNode.hs

```diff
@@ -299,20 +305,32 @@ node conf logger = do
     dbBaseDir <- getDbBaseDir conf
     when (_nodeConfigResetChainDbs conf) $ removeDirectoryRecursive dbBaseDir
     rocksDbDir <- getRocksDbDir conf
+    rocksDbCheckpointDir <- getRocksDbCheckpointDir conf
     pactDbDir <- getPactDbDir conf
     withRocksDb rocksDbDir $ \rocksDb -> do
         logFunctionText logger Info $ "opened rocksdb in directory " <> sshow rocksDbDir
-        withChainweb cwConf logger rocksDb pactDbDir (_nodeConfigResetChainDbs conf) $ \cw -> mapConcurrently_ id
-            [ runChainweb cw
-              -- we should probably push 'onReady' deeper here but this should be ok
-            , runCutMonitor (_chainwebLogger cw) (_cutResCutDb $ _chainwebCutResources cw)
-            , runQueueMonitor (_chainwebLogger cw) (_cutResCutDb $ _chainwebCutResources cw)
-            , runRtsMonitor (_chainwebLogger cw)
-            , runBlockUpdateMonitor (_chainwebLogger cw) (_cutResCutDb $ _chainwebCutResources cw)
-            ]
+        bracket (newIORef (Just rocksDb)) (flip writeIORef Nothing) $ \ref -> do
+            installHandlerCross sigUSR1 (const $ makeCheckpoint rocksDbCheckpointDir ref)
+            withChainweb cwConf logger rocksDb pactDbDir (_nodeConfigResetChainDbs conf) $ \cw -> mapConcurrently_ id
+                [ runChainweb cw
+                  -- we should probably push 'onReady' deeper here but this should be ok
+                , runCutMonitor (_chainwebLogger cw) (_cutResCutDb $ _chainwebCutResources cw)
+                , runQueueMonitor (_chainwebLogger cw) (_cutResCutDb $ _chainwebCutResources cw)
+                , runRtsMonitor (_chainwebLogger cw)
+                , runBlockUpdateMonitor (_chainwebLogger cw) (_cutResCutDb $ _chainwebCutResources cw)
+                ]
   where
     cwConf = _nodeConfigChainweb conf
 
+makeCheckpoint :: FilePath -> IORef (Maybe RocksDb) -> IO ()
+makeCheckpoint checkpointDir rdbRef = 
+    readIORef rdbRef >>= \case
+        Nothing -> pure ()
+        -- 0 ~ never flush WAL log before checkpoint, to avoid making extra work
+        Just rdb -> do 
+            now <- getCurrentTimeIntegral @Integer
+            checkpointRocksDb rdb 0 (checkpointDir </> formatTimeMicros now)
```

Comment:
I assume that the resulting file names lexicographically ordered, so that using `ls | tail -n 1` will provide the latest snapshot? Also, are all characters portable and compatible with all major file systems?

Reply:
Seems that it's ISO8601, for which lexicographic and chronologic orderings coincide. But... it's not compatible with NTFS or exFAT, which reserve `:` apparently. So I guess this should just be unix timestamps.

Reply:
hmm, not sure if we care about NTFS and FAT. Readable dates make snapshot maintenance easier. But scripts could still use the file creation time. So POSIX timestamps would be fine.

Not sure, I don't have a strong opinion here.